packages/edit-site/src/components/global-styles/preview-colors.js

```diff
@@ -0,0 +1,216 @@
+} from '@wordpress/components';
+import {
+	useThrottle,
+	useReducedMotion,
+	useResizeObserver,
+} from '@wordpress/compose';
+import { useLayoutEffect, useState, useMemo } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { unlock } from '../../lock-unlock';
+import { useStylesPreviewColors } from './hooks';
+
+const { useGlobalStyle, useGlobalStylesOutput } = unlock(
+	blockEditorPrivateApis
+);
+
+const firstFrame = {
+	start: {
+		scale: 1,
+		opacity: 1,
+	},
+	hover: {
+		scale: 0,
+		opacity: 0,
+	},
+};
+
+const normalizedWidth = 248;
+const normalizedHeight = 152;
+
+const normalizedColorSwatchSize = 66;
+
+// Throttle options for useThrottle. Must be defined outside of the component,
+// so that the object reference is the same on each render.
+const THROTTLE_OPTIONS = {
+	leading: true,
+	trailing: true,
+};
+
+const StylesPreviewColors = ( { label, isFocused, withHoverView } ) => {
+	const [ backgroundColor = 'white' ] = useGlobalStyle( 'color.background' );
+	const [ gradientValue ] = useGlobalStyle( 'color.gradient' );
+	const [ styles ] = useGlobalStylesOutput();
+	const disableMotion = useReducedMotion();
+	const [ isHovered, setIsHovered ] = useState( false );
+	const [ containerResizeListener, { width } ] = useResizeObserver();
+	const [ throttledWidth, setThrottledWidthState ] = useState( width );
+	const [ ratioState, setRatioState ] = useState();
+
+	const setThrottledWidth = useThrottle(
+		setThrottledWidthState,
+		250,
+		THROTTLE_OPTIONS
+	);
+
+	// Must use useLayoutEffect to avoid a flash of the iframe at the wrong
+	// size before the width is set.
+	useLayoutEffect( () => {
+		if ( width ) {
+			setThrottledWidth( width );
+		}
+	}, [ width, setThrottledWidth ] );
+
+	// Must use useLayoutEffect to avoid a flash of the iframe at the wrong
+	// size before the width is set.
+	useLayoutEffect( () => {
+		const newRatio = throttledWidth ? throttledWidth / normalizedWidth : 1;
+		const ratioDiff = newRatio - ( ratioState || 0 );
+
+		// Only update the ratio state if the difference is big enough
+		// or if the ratio state is not yet set. This is to avoid an
+		// endless loop of updates at particular viewport heights when the
+		// presence of a scrollbar causes the width to change slightly.
+		const isRatioDiffBigEnough = Math.abs( ratioDiff ) > 0.1;
+
+		if ( isRatioDiffBigEnough || ! ratioState ) {
+			setRatioState( newRatio );
+		}
+	}, [ throttledWidth, ratioState ] );
+
+	// Set a fallbackRatio to use before the throttled ratio has been set.
+	const fallbackRatio = width ? width / normalizedWidth : 1;
+	/*
+	 * Use the throttled ratio if it has been calculated, otherwise
+	 * use the fallback ratio. The throttled ratio is used to avoid
+	 * an endless loop of updates at particular viewport heights.
+	 * See: https://github.com/WordPress/gutenberg/issues/55112
+	 */
+	const ratio = ratioState ? ratioState : fallbackRatio;
+
+	const { highlightedColors } = useStylesPreviewColors();
```

Comment:
While testing I noticed an issue with some themes that had only 2 palette colors, e.g., Freddie.

If those colors match heading or background colors, they're being filtered out:

https://github.com/WordPress/gutenberg/blob/f7b735e9408c10f9600b1a0bcf2c2c0f1e827f80/packages/edit-site/src/components/global-styles/hooks.js#L68-L71

The result is that, although there's a variation, the colors are empty:

<img width="284" alt="Screenshot 2024-02-28 at 12 08 24 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/8ad8e200-2ae0-4d09-8394-6355e44c7416" />


CanvasLoader already gets around it this way:

https://github.com/WordPress/gutenberg/blob/f7b735e9408c10f9600b1a0bcf2c2c0f1e827f80/packages/edit-site/src/components/canvas-loader/index.js#L19-L23

Not sure what's appropriate in `<StylesPreviewColors />`

Maybe something like:

```js
	const { paletteColors, highlightedColors } = useStylesPreviewColors();

	const previewColors = highlightedColors.length
		? highlightedColors
		: paletteColors.slice( 0, 2 );
```

Reply:
Update: this is already an "feature" of trunk, in that style variation previews won't show them either. 

I think it just looks strange in the preset previews. Maybe good for an immediate follow up.

To replicate the above issue, here's a style variation example. The palette colors are the same as the background and text colors so the variation will show nothing.

```json
{
	"version": 2,
	"settings": {
	"color": {
		"palette": [
			{
				"color": "#000000",
				"name": "Base",
				"slug": "base"
			},
			{
				"color": "#ffffff",
				"name": "Contrast",
				"slug": "contrast"
			}
		]
	}
	},
	"styles": {
		"color": {
			"text": "#000000",
			"background": "var(--wp--preset--color--contrast)"
		}
	}
}

```